### PR TITLE
Support XLA replication and in/out axes in pmap

### PR DESCRIPTION
**Pull request by jekbradbury**

### CODE REVIEW

File: benchmarks/pmap_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+import jax
+import jax.numpy as np
+from jax import pmap
+from jax.util import safe_zip
+
+def pstd(x):
+  return x.std() / x.mean() * 100
+
+def benchmark(f, iters=None, warmup=None, name=None, target_total_secs=None):
```
Comment (gnecula) [inline benchmarks/pmap_benchmark.py]: It makes sense for this function to be in a shared module.

File: benchmarks/pmap_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  times = onp.array(times)
+  print("---------Benchmark results for %s---------" % (name or f.__name__))
+  print("mean=%f std=%f %%std=%f total=%f" %
+        (times.mean(), times.std(), pstd(times), times.sum()))
+  print("#iters=%d #warmup=%d" % (count, warmup))
+  print()
+  return times
+
+
+def benchmark_suite(funcs, params_list, param_names, name,
```
Comment (gnecula) [inline benchmarks/pmap_benchmark.py]: Why not pass params as a list of dictionaries? Maybe add also a short docstring?

File: benchmarks/pmap_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                    target_total_secs=None):
+  times = []
+  for f, params in safe_zip(funcs, params_list):
+    subname = name + "".join("_%s=%s" % (n, p)
+                             for n, p in safe_zip(param_names, params))
+    times.append(benchmark(f, name=subname,
```
Comment (gnecula) [inline benchmarks/pmap_benchmark.py]: I had expected here that f would be invoked with the params. I think this would be clearer if benchmark would be passed both a function and the parameters. To enable the staging of the input data, the benchmark function would be more like the get_benchmark_fn below (signature params -> unit -> unit). WDYT?

File: tests/pmap_test.py [hunk 4]
```diff
@@ -747,6 +749,25 @@ def testPswapaxes(self):
...
+    aval = ShapedArray((6,4), shard.dtype)
+    indices = [(slice(0,3), slice(0,2)),
+               (slice(0,3), slice(2,4)),
+               (slice(3,6), slice(0,2)),
+               (slice(3,6), slice(2,4))]
+    arr = pxla.ShardedDeviceArray(aval, indices, bufs)
+
+    r = pmap(lambda x: x + 1)(arr)
+    self.assertAllClose(r, arr + 1, check_dtypes=True)
+    self.assertEqual(len(r.device_buffers), 6)
```
Comment (gnecula) [inline tests/pmap_test.py]: I think we can also assert something about the indices of `r`. Perhaps instead of asserting the length of `r.device_buffers` assert the value of `d.logical_indices`. Then one can read from this test the sharding behavior.

File: jax/interpreters/ad.py [hunk 5]
```diff
@@ -556,10 +576,22 @@ def call_transpose(primitive, params, jaxpr, consts, freevar_vals, args, ct):
 
 def map_transpose(primitive, params, jaxpr, consts, freevar_vals, args, ct):
   all_args, in_tree_def = tree_flatten((consts, freevar_vals, args, ct))
+  # I think only one of `args` is non-undefined, but is only one of `ct` non-Zero?
+  # if that were the case we could use integers instead of tuples for the last two
+  # as is, it's unclear if it's even possible to create tree prefixes...
+  # instead, we'll want to separately flatten consts and freevars (using flatten_axes)
+  # and the other two
+  primal_dim = tuple(dim for arg, dim in zip(args, params['in_dims']) if arg is not undefined_primal)
+  in_dims = (0,) * len(tree_flatten((consts, freevar_vals))[0]) + \
```
Comment (gnecula) [inline jax/interpreters/ad.py]: All this will need to be revisited in light of the fact that `consts` and `freevar_vals` are now part of `args` :-(

Comment (gnecula) [reply jax/interpreters/ad.py]: There is also params['mapped_invars'] that tells you which `args` are mapped over.

File: jax/api_util.py [hunk 6]
```diff
@@ -70,3 +71,23 @@ def flatten_fun_nokwargs2(in_tree, *args_flat):
   ans_flat, ans_tree = tree_flatten(ans)
   aux_flat, aux_tree = tree_flatten(aux)
   yield (ans_flat, aux_flat), (ans_tree, aux_tree)
+
+def flatten_axes(treedef, axis_tree):
```
Comment (gnecula) [inline jax/api_util.py]: For functions like this I think that it is worth starting to add unit tests. We do most of our testing with integration tests and it is hard to know what kind of coverage we have. Also, unit tests can be helpful to understand in detail the behavior of this function.

File: jax/api_util.py [hunk 7]
```diff
@@ -14,7 +14,8 @@
 
 
 from .tree_util import (build_tree, tree_flatten, tree_unflatten,
-                        treedef_is_leaf)
+                        treedef_is_leaf, tree_multimap,
+                        _replace_nones)
```
Comment (gnecula) [inline jax/api_util.py]: Maybe remove the _ from `_replace_nones`?

File: jax/api.py [hunk 8]
```diff
@@ -652,71 +652,54 @@ def vmap(fun, in_axes=0, out_axes=0):
...
+                for size, idxs in sizes.items()]
+      raise ValueError(msg.format("\n".join(lines1 + ["so"] + lines2)))
+    else:
+      sizes = [x.shape[d] if d is not None else None for x, d in zip(vals, dims)]
+      sizes = tree_unflatten(tree, sizes)
+      raise ValueError(msg.format("the tree of axis sizes is:\n{}".format(sizes)))
 
 
-def pmap(fun, axis_name=None, devices=None, backend=None, axis_size=None):
+def pmap(fun, axis_name=None, in_axes=0, out_axes=0, devices=None, backend=None,
```
Comment (gnecula) [inline jax/api.py]: Please remember to change the CHANGELOG.md

File: jax/api.py [hunk 9]
```diff
@@ -769,13 +756,13 @@ def pmap(fun, axis_name=None, devices=None, backend=None, axis_size=None):
       Optional, a string representing the xla backend. 'cpu', 'gpu', or 'tpu'.
 
   Returns:
-    A parallelized version of ``fun`` with arguments that correspond to those of
-    ``fun`` but each with an additional leading array axis (with equal sizes)
-    and with output that has an additional leading array axis (with the same
-    size).
+    A parallelized version of ``fun`` with arguments that correspond to
+    those of ``fun``, but with extra array axes at positions indicated by
```
Comment (gnecula) [inline jax/api.py]: Perhaps this can be clarified: "...but each with an additional array axis at positions..."

File: jax/api.py [hunk 10]
```diff
@@ -769,13 +756,13 @@ def pmap(fun, axis_name=None, devices=None, backend=None, axis_size=None):
...
+    ``in_axes``, and a return value that corresponds to that of ``fun``, but
+    with extra array axes at positions indicated by ``out_axes``.
```
Comment (gnecula) [inline jax/api.py]: Suggest: ... but with each output having an extra array axis as positions indicated by ...

File: jax/api.py [hunk 11]
```diff
@@ -878,20 +865,23 @@ def pmap(fun, axis_name=None, devices=None, backend=None, axis_size=None):
...
-    out = pxla.xla_pmap(
+    out_flat = pxla.xla_pmap(
         flat_fun,
-        *args,
+        *args_flat,
         backend=backend,
         axis_name=axis_name,
         axis_size=local_axis_size,
         global_axis_size=axis_size,
+        in_dims=in_axes_flat,
```
Comment (gnecula) [inline jax/api.py]: I think that is makes sense to merge this "in_dims" with the newly added "mapped_invars" param. That one is at the moment only added deeper in the code for staged jaxprs, but has overlapping semantics.

File: jax/api.py [hunk 12]
```diff
@@ -596,7 +595,8 @@ def vmap(fun, in_axes=0, out_axes=0):
       container tree prefix of the positional argument tuple passed to ``fun``.
     out_axes: A nonnegative integer, None, or (nested) standard Python container
       (tuple/list/dict) thereof indicating where the mapped axis should appear
-      in the output.
+      in the output. Using None for a component of ``out_axes`` requires that
+      the corresponding value be identical across the "batch".
```
Comment (gnecula) [inline jax/api.py]: And we just take the first element?

**Comment by gnecula** [comment]
I had not noticed that this was a Draft PR, perhaps you did not intend to send it for review yet. Still, some of the comments may be helpful though.

**Comment by jekbradbury** [comment]
Yes, thanks! I definitely want to wait until Skye’s PR (that this is stacked atop) goes in, and I also need to rebase it past your changes (including folding in mapped_invars), but your comments are still helpful.

**Comment by hawkinsp** [comment]
Closing as stale.
